### PR TITLE
Cache feed

### PR DESCRIPTION
**Pull request by passaro**

### CODE REVIEW

File: mountpoint-s3/src/prefetch/cached_feed.rs [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+            }
+        }
+        trace!("request finished");
+    }
+
+    fn update_cache(&self, cache_key: &CacheKey, block_index: u64, block: &ChecksummedBytes) {
+        // TODO: consider updating the cache asynchronously
+        metrics::counter!("cache.total_bytes", block.len() as u64, "type" => "write");
+        match self.cache.put_block(cache_key.clone(), block_index, block.clone()) {
+            Ok(()) => {}
```
Comment (dannycjones) [inline mountpoint-s3/src/prefetch/cached_feed.rs:227]: might be worth timing the duration to write to cache

File: mountpoint-s3/src/prefetch/cached_feed.rs [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+    #[test_case(3 * MB, 8 * MB, 14 * MB, 0, 14 * MB; "whole object, size not aligned to parts or blocks")]
+    #[test_case(3 * MB, 8 * MB, 14 * MB, 9 * MB, 100 * MB; "aligned offset, size not aligned to parts or blocks")]
+    #[test_case(1 * MB, 8 * MB, 100 * KB, 0, 100 * KB; "small object")]
+    fn test_read_from_cache(
+        block_size: usize,
+        client_part_size: usize,
+        object_size: usize,
+        offset: usize,
+        preferred_size: usize,
+    ) {
```
Comment (dannycjones) [inline mountpoint-s3/src/prefetch/cached_feed.rs:297]: do we expect client part size to never be configured smaller than block size?

File: mountpoint-s3/src/prefetch/cached_feed.rs [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+        seed: u8,
+        part_queue: PartQueue<ObjectPartFeedError<impl std::error::Error + Send + Sync + 'static>>,
+    ) {
+        let mut offset = range.start();
+        let mut remaining = range.len();
+        while remaining > 0 {
+            let part = block_on(part_queue.read(remaining)).unwrap();
+            let bytes = part.into_bytes(key, offset).unwrap();
+
+            let expected = ramp_bytes(seed as usize + offset as usize, bytes.len());
```
Comment (dannycjones) [inline mountpoint-s3/src/prefetch/cached_feed.rs:401]: do we want some kind of `read_object` method on the mock client? I'd prefer we keep the byte ramp and seeds out of here

Comment (passaro) [reply mountpoint-s3/src/prefetch/cached_feed.rs:401]: Good point. I should have used MockObject actually

**Pull request closed without merging**